Account/FirefoxAccount.swift

```diff
@@ -190,13 +196,104 @@ open class FirefoxAccount {
+                    self.downloadAvatar()
+                    break
+                case .failedCanRetry:
+                    self.downloadAvatar()
+                    break
+                default:
+                    break
+                }
+            }
+            
+            func downloadAvatar() {
+                SDWebImageManager.shared().downloadImage(with: url, options: [SDWebImageOptions.continueInBackground, SDWebImageOptions.lowPriority], progress: nil) { (image, error, cacheType, success, url) in
+                    if let error = error {
+                        if (error as NSError).code == 404 || self.currentImageState == ImageDownloadState.failedCanRetry {
+                            // Image is not found or failed to download a second time
+                            self.currentImageState = ImageDownloadState.failedCanNotRetry
+                        } else {
+                            // This could have been a transient error, attempt to download the image only once more
+                            self.currentImageState = ImageDownloadState.failedCanRetry
+                            self.updateAvatarImageState()
+                        }
+                        return
+                    }
+                    
+                    if success == true && image == nil {
+                        self.currentImageState = ImageDownloadState.succeededMalformed
+                        return
+                    }
+                    
+                    self.image = image
+                    self.currentImageState = ImageDownloadState.succeeded
+                    
+                    // Delay posting notification by 100ms to avoid any racing observers
+                    DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + .milliseconds(100)) {
+                        NotificationCenter.default.post(name: NotificationFirefoxAccountProfileChanged, object: self)
+                    }
+                }
+            }
+        }
+    }
+    
+    // Fetch current user's FxA profile. It contains the most updated email, displayName and avatar. This
+    // emits a two `NotificationFirefoxAccountProfileChanged`, once when the profile has been downloaded and
+    // another when the avatar image has been downloaded.
+    open func updateProfile() {
+        guard let session = stateCache.value as? TokenState else {
+            return
+        }
+        
+        let client = FxAClient10(authEndpoint: self.configuration.authEndpointURL, oauthEndpoint: self.configuration.oauthEndpointURL, profileEndpoint: self.configuration.profileEndpointURL)
+        client.getProfile(withSessionToken: session.sessionToken as NSData) >>== { result in
+            self.fxaProfile = FxAProfile(email: result.email, displayName: result.displayName, avatar: result.avatarURL)
+            NotificationCenter.default.post(name: NotificationFirefoxAccountProfileChanged, object: self.fxaProfile)
```

Comment:
This notification is not posted on the main thread. You should be consistent about whether it is or not, and what observers need to do.